### PR TITLE
upstream: remove thread local cluster after triggering call backs

### LINKED ISSUE #7990: Envoy crashes with assertion failure on redis cluster removal
*Description*:
Envoy crashes with assertion failure if a redis cluster is removed.

- Management server sends redis cluster with CDS response
- Envoy is fully initialized
- Management server sends CDS update with out redis cluster
- Envoy tries to remove the cluster and crashes with the following assertion while removing membership update callback.

```
[2019-08-21 13:35:56.778][1462770][critical][assert] [bazel-out/darwin-fastbuild/bin/source/common/common/_virtual_includes/callback_impl_lib/common/common/callback_impl.h:70] assert failure: std::find_if(callbacks_.begin(), callbacks_.end(), [handle](const CallbackHolder& holder) -> bool { return handle == &holder; }) != callbacks_.end().
```
If I look at the `callbacks_.size() `here - it is actually zero.

https://github.com/envoyproxy/envoy/blob/master/source/common/common/callback_impl.h#L60

Any idea on callback is not present during removal?

### PR DESCRIPTION
**Pull request by ramaraochavali**
Description: Removes thread local cluster after triggering callbacks. Fixes #7990 
Risk Level: Low
Testing: Added Integration test
Docs Changes: N/A
Release Notes: N/A
Fixes #7990

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@mattklein123 @snowp I had to close the old PR (#7994 ) as github was not showing the changes pushed. 

I added an integration test in `ads_integration_test` that reproduces the issue and verified that fix actually works. PTAL.

File: test/integration/ads_integration.cc [hunk 1]
```diff
@@ -46,6 +46,17 @@ envoy::api::v2::Cluster AdsIntegrationTest::buildCluster(const std::string& name
...
 
+envoy::api::v2::Cluster AdsIntegrationTest::buildRedisCluster(const std::string& name) {
+  return TestUtility::parseYaml<envoy::api::v2::Cluster>(fmt::format(R"EOF(
+      name: {}
+      connect_timeout: 5s
+      type: EDS
+      eds_cluster_config: {{ eds_config: {{ ads: {{}} }} }}
+      lb_policy: MAGLEV
+    )EOF",
+                                                                     name));
```
Comment (snowp) [inline test/integration/ads_integration.cc]: formatting seems off here

Comment (ramaraochavali) [reply test/integration/ads_integration.cc]: It is same as buildCluster method, I thought it is some kept it like for some reason. Let me try formatting both.

Comment (ramaraochavali) [reply test/integration/ads_integration.cc]: If I change this and reformat, it is coming like this only. So may be formatter is like this.,

Comment (snowp) [reply test/integration/ads_integration.cc]: ok, that's fine

**Pull request merged**